Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1465 @@
+Authors: María Inés de Frutos Fernández, Xavier Généreux
+-/
+import Mathlib.Algebra.Algebra.Equiv
+import Mathlib.Algebra.Algebra.NonUnitalHom
+import Mathlib.Algebra.Module.BigOperators
+import Mathlib.Data.Finsupp.Basic
+
+/-!
+
+This file presents a skewed version of `Mathlib.Algebra.MonoidAlgebra.Basic`.
+
+We define `SkewMonoidAlgebra k G` attached with a skewed convolution product.
+Here, the product of two elements `f g : SkewMonoidAlgebra k G` is the finitely supported
+function whose value at `a` is the sum of `f x * (x • g y)` over all pairs `x, y`
+such that `x * y = a`.
+
+The associativity of the skewed multiplication depends on the `[MulSemiringAction G k]` instance.
+In particular, this means that unlike in `Mathlib.Algebra.MonoidAlgebra.Basic`, `G` will
+need to be a monoid for most of our uses.
+-/
+
+noncomputable section
+
+/-- The skew monoid algebra over a semiring `k` generated by the monoid `G`.
+It is the type of finite formal `k`-linear combinations of terms of `G`,
+endowed with a skewed convolution product.
+-/
+@[ext]
+structure SkewMonoidAlgebra (k : Type*) (G : Type*) [Zero k] where
+  /-- Map **from** `G →₀ k`. -/
+  ofFinsupp ::
+  /-- Map **to** `G →₀ k`. -/
+  toFinsupp : G →₀ k
+
+open Function hiding Commute
+namespace SkewMonoidAlgebra
+
+universe u₁ u₂
+
+variable {k : Type u₁} {G : Type u₂} {H : Type*} {R : Type*}
+
+section AddCommMonoid
+
+variable [AddCommMonoid k]
+
+@[simp]
+theorem eta (f : SkewMonoidAlgebra k G) : ofFinsupp f.toFinsupp = f := by cases f; rfl
+
+private irreducible_def add :
+    SkewMonoidAlgebra k G → SkewMonoidAlgebra k G → SkewMonoidAlgebra k G
+  | ⟨a⟩, ⟨b⟩ => ⟨a + b⟩
+
+instance zero : Zero (SkewMonoidAlgebra k G) := ⟨⟨0⟩⟩
+
+instance add' : Add (SkewMonoidAlgebra k G) := ⟨add⟩
+
+instance smulZeroClass {S : Type*} [SMulZeroClass S k] :
+    SMulZeroClass S (SkewMonoidAlgebra k G) where
+  smul r p := ⟨r • p.toFinsupp⟩
+  smul_zero a := congr_arg ofFinsupp (smul_zero a)
+
+@[simp]
+theorem ofFinsupp_zero : (⟨0⟩ : SkewMonoidAlgebra k G) = 0 := rfl
+
+@[simp]
+theorem ofFinsupp_add {a b} : (⟨a + b⟩ : SkewMonoidAlgebra k G) = ⟨a⟩ + ⟨b⟩ :=
+  show _ = add _ _ by rw [add_def]
+
+@[simp]
+theorem ofFinsupp_smul {S : Type*} [SMulZeroClass S k] (a : S) (b : G →₀ k) :
+    (⟨a • b⟩ : SkewMonoidAlgebra k G) = (a • ⟨b⟩ : SkewMonoidAlgebra k G) :=
+  rfl
+
+@[simp]
+theorem toFinsupp_zero : (0 : SkewMonoidAlgebra k G).toFinsupp = 0 := rfl
+
+@[simp]
+theorem toFinsupp_add (a b : SkewMonoidAlgebra k G) :
+    (a + b).toFinsupp = a.toFinsupp + b.toFinsupp := by
+  cases a
+  cases b
+  rw [← ofFinsupp_add]
+
+@[simp]
+theorem toFinsupp_smul {S : Type*} [SMulZeroClass S k] (a : S) (b : SkewMonoidAlgebra k G) :
+    (a • b).toFinsupp = a • b.toFinsupp :=
+  rfl
+
+theorem _root_.IsSMulRegular.skewMonoidAlgebra {S : Type*} [Monoid S] [DistribMulAction S k] {a : S}
+    (ha : IsSMulRegular k a) : IsSMulRegular (SkewMonoidAlgebra k G) a
```

Comment:
This could be an `iff`, but I don't know whether it's important to state formally (the converse holds in large generality…)

Reply:
As I said above, the lemma should be about free (nonzero — you're right) modules. I'll be happy if it's in the TODO list.